core/src/main/scala/kafka/server/MetadataCache.scala

```diff
@@ -356,6 +353,84 @@ class MetadataCache(brokerId: Int) extends Logging {
     }
   }
 
+  private def logListenersNotIdenticalIfNecessary(aliveNodes: mutable.LongMap[collection.Map[ListenerName, Node]]) = {
+    aliveNodes.get(brokerId).foreach { listenerMap =>
+      val listeners = listenerMap.keySet
+      if (!aliveNodes.values.forall(_.keySet == listeners))
+        error(s"Listeners are not identical across brokers: $aliveNodes")
+    }
+  }
+
+  private def different(brokerRecord: BrokerRecord, broker: Broker): Boolean = {
+    if (brokerRecord.brokerId() != broker.id) {
+      return true
+    }
+    val brokerRecordRack = brokerRecord.rack()
+    // not sure if "" may imply no rack since BrokerRecord default constructor sets this.rack=""
+    val brokerRecordHasRack = brokerRecordRack != null && brokerRecordRack != ""
+    if (!brokerRecordHasRack && broker.rack.isDefined || brokerRecordHasRack && broker.rack.isEmpty) {
+      return true
+    }
+    broker.rack.foreach( rack => if (rack != brokerRecordRack) return true)
+    // We are going to have to compare sizes no matter what, so do that first even if this Seq implementation is O(n)
+    if (broker.endPoints.size != brokerRecord.endPoints().size()) {
+      return true
+    }
+    broker.endPoints.foreach(brokerEndPoint => {
+      val brokerRecordEndpoint = brokerRecord.endPoints().find(brokerEndPoint.listenerName.value())
+      if (brokerRecordEndpoint == null) {
+        return true
+      }
+      if (brokerEndPoint.host != brokerRecordEndpoint.host() ||
+        brokerEndPoint.port != brokerRecordEndpoint.port() ||
+        brokerEndPoint.securityProtocol.id != brokerRecordEndpoint.securityProtocol()) {
+        return true
+      }
+    })
+    false
+  }
+
+  def updateMetadata(brokerRecord: BrokerRecord): Unit = {
+    inWriteLock(partitionMetadataLock) {
+      /*
+      * Upsert the single given broker while doing nothing if there is no change
```

Comment:
We don't need logic to determine if there was a change or not.  Since the metadata log is ordered, we can just apply everything in the log.